### PR TITLE
Add PaymentRequest.prototype.hasEnrolledInstrument()

### LINKED ISSUE #830: Spec hasEnrolledInstrument
I started implementing `hasEnrolledInstrument` in Chrome according to the WG agreement [1]. It will be great if this API and the change to `canMakePayment` can be added to the spec language. I'd like to kick off an Intent to Ship thread in blink-dev@ soon as an official announcement of the `canMakePayment` behavior change to web developers.

[1] https://lists.w3.org/Archives/Public/public-payments-wg/2018Nov/0018.html

### PR DESCRIPTION
**Pull request by danyao**
closes #830 

The following tasks have been completed:

 * [x] Confirmed there are no ReSpec errors/warnings.
 * [x] Modified Web platform tests: https://github.com/web-platform-tests/wpt/pull/15306
 * [ ] Modified MDN Docs (link)
 * [ ] Has undergone security/privacy review (link)
 
Implementation commitment:

 * [x] [Safari](https://bugs.webkit.org/show_bug.cgi?id=197386)
 * [x] [Chrome](https://crbug.com/915907)
 * [ ] [Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1528663)
 * [ ] Edge (public signal)

Optional, impact on Payment Handler spec?
After this change, `CanMakePaymentEvent` will not be triggered on any payment handler when `canMakePayment()` runs. It will be triggered when `hasEnrolledInstrument()` runs.

***
<a href="https://pr-preview.s3.amazonaws.com/danyao/payment-request/pull/833.html" title="Last updated on May 3, 2019, 2:50 AM UTC (6faac81)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/payment-request/833/72de3ba...danyao:6faac81.html" title="Last updated on May 3, 2019, 2:50 AM UTC (6faac81)">Diff</a>

### CODE REVIEW

**Comment by marcoscaceres** [comment]
Thanks @danyao! I’ll try to review this ASAP!

**Comment by danyao** [comment]
Thanks! I forgot to update the PaymentRequest interface with the new method in the first commit. This is fixed now.

File: index.html
Comment (ianbjacobs) [inline index.html]:
```suggestion
              <li>If the user agent has a <a>payment handler</a> that supports
```

File: index.html
Comment (ianbjacobs) [inline index.html]:
```suggestion
          e.g., when showing a "buy now" button). If the method is called too
```

File: index.html
Comment (ianbjacobs) [inline index.html]:
```suggestion
          often, the user agent may return <a>a promise rejected
```

File: index.html
Comment (ianbjacobs) [inline index.html]: I have a question here: Instead of "steps to check if a payment can be made" should this be "steps to check if the handler has an enrolled instrument"? And does this mean that a payment method also needs to define those steps? Does a revised basic card spec need to change the steps to be about enrolled instrument?

(I don't have answers to these questions yet.)

[1] https://www.w3.org/TR/payment-method-basic-card/#steps-to-check-if-a-payment-can-be-made

**Review by ianbjacobs — DISMISSED** [review]

File: index.html
Comment (danyao) [reply index.html]: In Chrome we're using the existing steps in the basic card spec to determine if a card should be considered "enrolled". For payment handler, I believe the equivalent is the CanMakePaymentEvent. In both cases, our rationale is that the new canMakePayment() method checks the existence of a payment handler for the specified method, and hasEnrolledInstrument() checks the actual "can-make-payment" property of the payment handler.

I think renaming "steps to check if a payment can be made" to "steps to check if the handler has an enrolled instrument" may be more clear in the long run, but it introduces a lot of changes to existing spec without changing any behavior. That's why I opted to reuse the existing language as much as possible.

File: index.html
Comment (marcoscaceres) [inline index.html]:
```suggestion
        <div class="note" title="canMakePayment() vs hasEnrolledInstrument()">
```

File: index.html
Comment (marcoscaceres) [inline index.html:1318]: I realize now that the first sentence of this paragraphs is somewhat misleading, but the second part is definitely correct. I'd suggest:

```HTML
          <p>
            The <a>canMakePayment()</a> method returns a <a>Promise</a>
            that is fulfilled with true if the <a>user agent</a> supports
            any of the <a data-link-for="PaymentMethodData">supportedMethods</a> supplied to the
            <a>PaymentRequest</a> <a data-lt="PaymentRequest.PaymentRequest()">constructor</a>, or false if none are supported.
          </p>
          <p>
            A true result from <a>canMakePayment()</a> does not imply that the
            user has an provisioned instrument ready for payment. For that, use
            <a>hasEnrolledInstrument()</a> instead.
          </p>
```

Needs a tidy :)

File: index.html
Comment (marcoscaceres) [inline index.html]:
```suggestion
          the desired <a>payment methods</a> but is also "ready for payment" (
```

File: index.html
Comment (marcoscaceres) [inline index.html]:
```suggestion
          with </a> a "<a>NotAllowedError</a>" <a>DOMException</a> at its
```

**Review by marcoscaceres — CHANGES_REQUESTED** [review]
Hi @danyao, this is a great start - and the algorithm parts are looking pretty solid... 

However, I'd like to do this in two parts because there is some duplication in the two algorithms that I'd like to avoid (mostly an editorial/maintenance issue). Also note that the `hasEnrolledInstrument()` doesn't seem to integrate with the state machine like the other methods.

For example, right now:

```
pr = new PaymentRequest("")
await pr.abort();
await pr.canMakePayment(); //throws
await pr.hasEnrolledInstrument(); // doesn't throw.
``` 

I think we should have a single algorithm in the Algorithms sections. The algorithm should be called:

"can make payment algorithm" and have it take an argument Boolean `<var>checkForInstruments</var>` that if passed as "true", it does the instrument check.   

Then,  instead of:

```HTML
The <a>canMakePayment()</a> method MUST act as follows:
```

It would be:

```
The <a>canMakePayment()</a> method MUST run the <a>can make payment algorithm</a> with the <var>checkForInstruments</a> set to false.
```

Then,  instead of:

```HTML
The <a>hasEnrolledInstrument()</a> method MUST act as follows:
```

It would be:

```HTML
The <a>hasEnrolledInstrument()</a> method MUST run the <a>can make payment algorithm</a> with the <var>checkForInstruments</a> set to true.
```

**Comment by marcoscaceres** [comment]
Now I'm wondering if we should just add a optional dictionary to `canMakePayment()`:

```JS
pr.canMakePayment({ withEnrolledInstruments: true });
```

**Comment by webpayments-specs** [comment]
That would make feature detection difficult, so I would advise against it.

**Comment by marcoscaceres** [comment]
Firefox would likely never return this information truthfully for built in methods regardless (i.e., we will always return true), so I don’t see what difference it makes from a feature detection POV?

**Comment by marcoscaceres** [comment]
(Note the hasEnrolledInstruments() would introduce another fingerprinting vector - so we might want to document that in the Privacy and Security section, and we should make it ok for UAs to lie for privacy reasons).

**Comment by rsolomakhin** [comment]
Chrome would return the `hasEnrolledInstrument()` truthfully for the built-in `basic-card` payment method. The feature detection matters for Chrome, because the old implementation of `canMakePayment()` was checking for instruments, whereas the new implementation would not be doing that. The only way for a website to determine which is which is to check for `hasEnrolledInstrument()` in the case of Chrome. Sorry about our legacy shipped code! :-(

> Note the hasEnrolledInstruments() would introduce another fingerprinting vector - so we might want to document that in the Privacy and Security section.

Privacy and Security section changes sound reasonable.

> We should make it ok for UAs to lie for privacy reasons.

Did you have in mind returning `NotAllowedError` similar to `canMakePayment()`? That sounds OK.

**Comment by marcoscaceres** [comment]
Ok, let’s go with the new name 👍 about the exceptions, see my suggestion about using the same abstract algorithm for both this an canMakePayment(). Then they are both protected by the same abuse checking.

File: index.html
Comment (danyao) [reply index.html:1318]: Done!

**Comment by danyao** [comment]
Hi @marcoscaceres, thanks for the feedback!

> However, I'd like to do this in two parts because there is some duplication in the two algorithms that I'd like to avoid (mostly an editorial/maintenance issue).

Consolidating redundant algorithm makes sense. I went the other way initially because I felt it may be confusing to embed an "if" and the rate-limit note inside the loop. I gave it another try. PTAL.

> 
 Also note that the `hasEnrolledInstrument()` doesn't seem to integrate with the state machine like the other methods.
> 
> For example, right now:
> 
> ```
> pr = new PaymentRequest("")
> await pr.abort();
> await pr.canMakePayment(); //throws
> await pr.hasEnrolledInstrument(); // doesn't throw.
> ```

This would be a bug. How did you test it? I added a manual WPT test case for this scenario and it was passing on my debug build, but I could have overlooked something: https://github.com/web-platform-tests/wpt/blob/master/payment-request/payment-request-hasenrolledinstrument-method-manual.https.html#L81

> (Note the hasEnrolledInstruments() would introduce another fingerprinting vector - so we might want to document that in the Privacy and Security section, and we should make it ok for UAs to lie for privacy reasons).

Updated the Privacy Considerations section. PTAL.

Do you think it's important to keep the abuse protection for `canMakePayment()`? I feel it's no longer necessary because with the introduction of JIT and removal of handler-specific can make payment check, the new `canMakePayment()` is mostly an UA version detector. `hasEnrolledInstrument()` on the other hand still exposes user information, so should be kept under abuse protection.

**Comment by danyao** [comment]
Friendly ping @marcoscaceres. I've responded to your feedback. PTAL.

File: index.html
Comment (marcoscaceres) [inline index.html]:
```suggestion
          specifies whether the algorithm checks for existence of enrolled
```

File: index.html
Comment (marcoscaceres) [inline index.html]:
```suggestion
          instruments in addition to supporting <a>payment method</a>.
```

File: index.html
Comment (marcoscaceres) [inline index.html]:
```suggestion
              abuse of the calling method for
```

File: index.html
Comment (marcoscaceres) [inline index.html]: Note that canMakePayment() is also a fingerprinting vector, so it’s prudent for this to apply to both.

File: index.html
Comment (marcoscaceres) [inline index.html]: I don't think we should mention the "just-in-time" feature... that's a Chrome implementation detail that falls under "supports".

File: index.html
Comment (marcoscaceres) [inline index.html]:
```suggestion
          agent</a> supports making payment with the <a>payment methods</a> with which the <a>PaymentRequest</a> was constructed. It
```

File: index.html
Comment (marcoscaceres) [inline index.html]:
```suggestion
            user has a provisioned instrument ready for payment. For that, use
```

File: index.html
Comment (marcoscaceres) [inline index.html]:
```suggestion
          <code>hasEnrolledInstrument()</code> protections
```

File: index.html
Comment (marcoscaceres) [inline index.html]:
```suggestion
       <section id="hasEnrolledInstrument-privacy">
```

File: index.html
Comment (danyao) [reply index.html]: Done.

File: index.html
Comment (danyao) [reply index.html]: Makes sense. Removed the JIT language.

**Comment by danyao** [comment]
@marcoscaceres All comments addressed. PTAL.

File: index.html
Comment (marcoscaceres) [inline index.html]: I'd suggest dropping the first sentence as it redefines what was already stated in the method's definition. Instead, just say something like: 

"The <a>hasEnrolledInstrument()</a> method has the potential to expose user information that could be abused for fingerprinting purposes. The API allows the user agent to restrict the number or frequency of calls to reduce the risk of fingerprinting. User agents MAY allow the user to control the response to <a>hasEnrolledInstrument()</a> via UI or provide their own means to protection (e.g., always returning with true)."

File: index.html
Comment (marcoscaceres) [inline index.html]: I'd suggest dropping everything after "but also..." and replacing it with:

"and if a <a>payment handler</a> has an instrument ready for payment. See <a href="#"hasEnrolledInstrument-privacy"><a> for privacy considerations."

And also drop the actual details (i.e., the "If the method is called too often...").

**Review by marcoscaceres — CHANGES_REQUESTED** [review]
Getting there... there is still a bit or repetition and some other little bits that I'd like to suggest we change.

File: index.html
Comment (marcoscaceres) [inline index.html]:
```suggestion
          <a>DOMException</a>.
```

File: index.html
Comment (marcoscaceres) [inline index.html]: I think we can drop this `li`, because we did this check in the constructor so this can't ever throw.

File: index.html
Comment (marcoscaceres) [inline index.html]:
```suggestion
          supporting a <a>payment method</a>.
```

File: index.html
Comment (danyao) [reply index.html]: Done.

File: index.html
Comment (danyao) [reply index.html]: Done. I also updated the corresponding note in canMakePayment() section above to match the wording, and also refer to the privacy section.

File: index.html
Comment (danyao) [reply index.html]: This is better! Thanks for the suggestion.

File: index.html
Comment (danyao) [reply index.html]: Done.

**Comment by marcoscaceres** [comment]
Thanks for the speedy turn around. Will give it one more look over, but looking really good.

**Review by marcoscaceres — APPROVED** [review]
LGTM. Thanks for doing this @danyao.  

Before we merge, we need:
 1. Approval and implementation commitment from @aestes. 
 1. Firefox bug (I'll file one).

We also need to decide if we are going to fork 1.1 at this point. I'm waiting on final review feedback from various Editors of 1.0.

**Comment by marcoscaceres** [comment]
Firefox bug https://bugzilla.mozilla.org/show_bug.cgi?id=1528663

**Comment by danyao** [comment]
Friendly ping, @aestes.

File: index.html
Comment (aestes) [inline index.html]:
```suggestion
          method via UI or provide their own means of protection (e.g., always
```

**Review by aestes — APPROVED** [review]

File: index.html
Comment (danyao) [reply index.html]: Hi @ianbjacobs - did I answer your question? This is the last open review question blocking this PR. :-)

**Comment by danyao** [comment]
Hi @marcoscaceres,

Now that this PR is approved, what would be the best next step? Should we create a 1.1 branch to commit this?

File: index.html
Comment (ianbjacobs) [reply index.html]: @danyao, sorry for the delay. Thank you for the answer. I'm ok with your proposal.

**Comment by ianbjacobs** [comment]
@plehegar do you have guidance here for introducing a new branch?

Ian

**Comment by marcoscaceres** [comment]
@danyao @plehegar @ianbjacobs, my preference is for us not to spin up new branches as we are almost ready to republish in the next 2 weeks. 

@danyao, just hang tight, we are almost done with 1.0. Once the new CR is out, we don't expect any significant feedback - so it will be easier to deal with back porting, or uplifting, changes as needed (and all this stuff will go into the `gh-pages` branch).

**Comment by marcoscaceres** [comment]
If we can get commitment from one more implementer we can land this.

**Comment by ianbjacobs** [comment]
@danyao, @rsolomakhin, @zouhir, @aestes see above request from @marcoscaceres. :)

**Comment by rsolomakhin** [comment]
@ianbjacobs : This is [shipped in Chrome](https://crbug.com/915907) and there's a [bug in WebKit](https://bugs.webkit.org/show_bug.cgi?id=197386). Is that sufficient?

**Comment by marcoscaceres** [comment]
@rsolomakhin, excellent! thanks for doing the archeology! Yep, that's good enough.

**Comment by ianbjacobs** [comment]
That sounds like 2 implementation commitments. @marcoscaceres, seem ok?

**Comment by marcoscaceres** [comment]
@danyao could you please check the merge conflict? I can modernize the markup a bit after we merge.

**Comment by danyao** [comment]
Yep I'll take a pass later today. Thanks for keeping on top of this!

**Comment by danyao** [comment]
Hi @marcoscaceres - I fixed the merge conflicts. PTAL!

**Comment by marcoscaceres** [comment]
Thanks @danyao! looks great.

**Pull request merged**
